### ISSUE TITLE
OpenCL FFT Speeds

### ISSUE THREAD

**Issue by whiteway11890** [arrayfire/arrayfire#2052]
I am working on a project that requires the use of many large FFTs. My group is targeting using the OpenCL backend for CPU. We have noticed the the ArrayFire times for FFTs are very slow for both the CPU and OpenCL backends. I ran quite a few tests and the numbers back up my observations.

I wrote a quick tester to look into the issue. I compiled it with icpc and limited the number of runtime threads using `taskset -c x ./fftPerformanceTester` where x is 0, 0-1, 0-3, or 0-7, respectively, for the chart below. I'm more than happy to furnish any other information that would help. It's relatively high priority to have a fast FFT for our purposes, and this unfortunately may keep us from moving more of our stuff to use arrayfire.

Please let me know if there is anything else I can do to help.

I ran my tester [fftTest.gz](https://github.com/arrayfire/arrayfire/files/1704371/fftTest.gz) on two different machines and here are the results that I came up with:
# CPU 1
```
CPU Info:
ArrayFire v3.6.0 (CPU, 64-bit Linux, build 267302d)
[0] Intel: Intel(R) Xeon(R) CPU E5-2697 v4 @ 2.30GHz, 193195 MB, Max threads(36)

Cuda Info:
ArrayFire v3.6.0 (CUDA, 64-bit Linux, build 267302d)
Platform: CUDA Toolkit 8, Driver: 384.98
[0] Quadro M5000, 8121 MB, CUDA Compute 5.2

OpenCL Info:
ArrayFire v3.6.0 (OpenCL, 64-bit Linux, build 267302d)
[0] INTEL: Intel(R) Xeon(R) CPU E5-2697 v4 @ 2.30GHz, 193195 MB

+------+---------+--------------+-----------+-----------+-----------+-----------+
| Type | Threads | Size         | CPU Time  | Cuda Time | OCL Time  | FFTW Time |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       1 |  32x 32x1000 |  0.004673 |  0.000333 |  0.016731 |  0.003557 |
|  C64 |       1 |  32x 32x1000 |  0.006772 |  0.001048 |  0.019762 |  0.005442 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       1 |  64x 64x1000 |  0.022485 |  0.001233 |  0.174248 |  0.017971 |
|  C64 |       1 |  64x 64x1000 |  0.054008 |  0.004544 |  0.260335 |  0.045586 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       1 | 128x128x1000 |  0.149270 |  0.004389 |  1.463000 |  0.122869 |
|  C64 |       1 | 128x128x1000 |  0.282295 |  0.019264 |  0.910718 |  0.233269 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       2 |  32x 32x1000 |  0.005229 |  0.000328 |  0.010561 |  0.003449 |
|  C64 |       2 |  32x 32x1000 |  0.015585 |  0.001043 |  0.012437 |  0.004346 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       2 |  64x 64x1000 |  0.025560 |  0.001239 |  0.088354 |  0.009622 |
|  C64 |       2 |  64x 64x1000 |  0.044249 |  0.004528 |  0.131070 |  0.034420 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       2 | 128x128x1000 |  0.092536 |  0.004324 |  0.734155 |  0.091354 |
|  C64 |       2 | 128x128x1000 |  0.161849 |  0.019258 |  0.477848 |  0.142999 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       4 |  32x 32x1000 |  0.003651 |  0.000340 |  0.006496 |  0.003061 |
|  C64 |       4 |  32x 32x1000 |  0.011069 |  0.001032 |  0.007637 |  0.002655 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       4 |  64x 64x1000 |  0.027884 |  0.001113 |  0.050602 |  0.005169 |
|  C64 |       4 |  64x 64x1000 |  0.046108 |  0.004532 |  0.066548 |  0.010907 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       4 | 128x128x1000 |  0.094211 |  0.004382 |  0.368076 |  0.049251 |
|  C64 |       4 | 128x128x1000 |  0.104709 |  0.019265 |  0.238845 |  0.083137 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       8 |  32x 32x1000 |  0.003214 |  0.000329 |  0.003984 |  0.002898 |
|  C64 |       8 |  32x 32x1000 |  0.004384 |  0.001042 |  0.004658 |  0.001856 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       8 |  64x 64x1000 |  0.014327 |  0.001241 |  0.029369 |  0.003724 |
|  C64 |       8 |  64x 64x1000 |  0.015035 |  0.004538 |  0.039634 |  0.006293 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       8 | 128x128x1000 |  0.070512 |  0.004396 |  0.184058 |  0.025945 |
|  C64 |       8 | 128x128x1000 |  0.078621 |  0.019269 |  0.118315 |  0.045322 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
```

# CPU 2

```
CPU Info:
ArrayFire v3.4.2 (CPU, 64-bit Linux, build 2da9967)
[0] Intel: Intel(R) Xeon(R) CPU E5-2643 0 @ 3.30GHz, 64231 MB, Max threads(8)

Cuda Info:
ArrayFire v3.4.2 (CUDA, 64-bit Linux, build 2da9967)
Platform: CUDA Toolkit 8, Driver: 375.66
[0] Tesla K20m, 5062 MB, CUDA Compute 3.5
-1- Tesla K20m, 5062 MB, CUDA Compute 3.5

OpenCL Info:
ArrayFire v3.4.2 (OpenCL, 64-bit Linux, build 2da9967)
[0] INTEL   : Intel(R) Xeon(R) CPU E5-2643 0 @ 3.30GHz, 64231 MB

+------+---------+--------------+-----------+-----------+-----------+-----------+
| Type | Threads | Size         | CPU Time  | Cuda Time | OCL Time  | FFTW Time |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       1 |  32x 32x1000 |  0.005130 |  0.000286 |  0.015276 |  0.002985 |
|  C64 |       1 |  32x 32x1000 |  0.008052 |  0.000534 |  0.018272 |  0.004642 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       1 |  64x 64x1000 |  0.025102 |  0.000987 |  0.120588 |  0.016074 |
|  C64 |       1 |  64x 64x1000 |  0.043745 |  0.001941 |  0.173415 |  0.029991 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       1 | 128x128x1000 |  0.128316 |  0.003837 |  0.465736 |  0.100223 |
|  C64 |       1 | 128x128x1000 |  0.252249 |  0.007841 |  0.696664 |  0.179157 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       2 |  32x 32x1000 |  0.003719 |  0.000293 |  0.015573 |  0.001781 |
|  C64 |       2 |  32x 32x1000 |  0.005911 |  0.000536 |  0.018545 |  0.002430 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       2 |  64x 64x1000 |  0.036544 |  0.000987 |  0.120752 |  0.008142 |
|  C64 |       2 |  64x 64x1000 |  0.029142 |  0.001945 |  0.111281 |  0.015135 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       2 | 128x128x1000 |  0.084748 |  0.003844 |  0.235290 |  0.051693 |
|  C64 |       2 | 128x128x1000 |  0.161930 |  0.007847 |  0.363072 |  0.083229 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       4 |  32x 32x1000 |  0.003042 |  0.000289 |  0.004247 |  0.001676 |
|  C64 |       4 |  32x 32x1000 |  0.005951 |  0.000528 |  0.007144 |  0.001511 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       4 |  64x 64x1000 |  0.020220 |  0.000982 |  0.032805 |  0.004404 |
|  C64 |       4 |  64x 64x1000 |  0.035207 |  0.001939 |  0.047348 |  0.008123 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       4 | 128x128x1000 |  0.118165 |  0.003837 |  0.126397 |  0.024313 |
|  C64 |       4 | 128x128x1000 |  0.098697 |  0.007885 |  0.201822 |  0.051564 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       8 |  32x 32x1000 |  0.003365 |  0.000298 |  0.002527 |  0.003111 |
|  C64 |       8 |  32x 32x1000 |  0.004453 |  0.000541 |  0.003480 |  0.001737 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       8 |  64x 64x1000 |  0.013609 |  0.000982 |  0.018975 |  0.015180 |
|  C64 |       8 |  64x 64x1000 |  0.031006 |  0.001945 |  0.024855 |  0.008441 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       8 | 128x128x1000 |  0.090327 |  0.003832 |  0.064745 |  0.024268 |
|  C64 |       8 | 128x128x1000 |  0.103075 |  0.007846 |  0.107863 |  0.043847 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
```

**Comment by pavanky** [comment]
OpencCL is not optimized for FFT on CPU devices. This situation may change if https://github.com/arrayfire/arrayfire/issues/2049 is resolved.

As for the benchmark, you need to change initialization to be like this for accurate results:

```
af::array fftData = af::constant(0, fftSize, type);
af::eval(fftData);
af::sync();
```

**Comment by pavanky** [comment]
You are also running the benchamrk for *1* iteration. This will not give you completely reliable numbers!

**Comment by whiteway11890** [comment]
The for loop in the code runs each function twice to get past the JIT compilation times:
```
//OPENCL
af::setBackend(AF_BACKEND_OPENCL);
times[2] = runFFT(fftSize, dimType);
times[2] = runFFT(fftSize, dimType);
```

I will change the initialization to what you have suggested and rerun my numbers.

**Comment by whiteway11890** [comment]
I won't flood the comments with all of the results I posted above, but here are the times after adding the eval and sync

```
+------+---------+--------------+-----------+-----------+-----------+-----------+
| Type | Threads | Size         | CPU Time  | Cuda Time | OCL Time  | FFTW Time |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       8 |  32x 32x1000 |  0.001142 |  0.000266 |  0.004003 |  0.003094 |
|  C64 |       8 |  32x 32x1000 |  0.002559 |  0.000943 |  0.004547 |  0.001645 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       8 |  64x 64x1000 |  0.012718 |  0.001038 |  0.028413 |  0.003970 |
|  C64 |       8 |  64x 64x1000 |  0.011957 |  0.004151 |  0.038785 |  0.009905 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
|  C32 |       8 | 128x128x1000 |  0.033973 |  0.003614 |  0.180043 |  0.026710 |
|  C64 |       8 | 128x128x1000 |  0.030255 |  0.017688 |  0.114448 |  0.046729 |
+------+---------+--------------+-----------+-----------+-----------+-----------+
```

**Comment by whiteway11890** [comment]
I could possibly contribute to #2049 if there is a plan to do it. That would help us out immensely. Would that basically constitute falling back to fftw for OpenCL CPU backends?

**Comment by pavanky** [comment]
@whiteway11890 yes. If you look at blas.cpp, it falls back to MKL for CPUs. You can do something similar for ffts in OpenCL as well.

Also the new numbers dont entirely make sense :-/ Arrayfire is faster for C64 as opposed to C32 for the larger sizes. It is also faster than fftw at the largest size..

**Comment by whiteway11890** [comment]
I'll have a look at how you implemented the BLAS fallback. We may be willing to write an FFT equivalent and contribute it.

**Comment by christophe-murphy** [comment]
This will be resolved when addressing #2049 , closing
